Dependency refactoring

petersteneteg opened the pull request:
Added VCPKG support for all dependencies except glad, (catch2 is also not added) 
Also added a vcpkg.json file to make it easy to use with vcpkg.  And a CMakePresets.json file for common setups. The  vcpkg preset assumes that the vcpkg repo lives next to the sgct repo.

Also fixes a few errors in the examples apps

petersteneteg committed:
Dependency refactoring

Added VCPKG support for all dependencies except glad, (catch2 is also not added)
Also added a vcpkg.json file to make it easy to use with vcpkg. and  a CMakePresets.json file for common setups. The  vcpkg preset assumes that the vcpkg repo lives next to the sgct repo.

alexanderbock commented on CMakeLists.txt:
```diff
@@ -44,17 +45,19 @@ if (WIN32)
 endif ()
 
 # Exceptions for external libraries
-option(SGCT_DEP_INCLUDE_FMT "Include FMT library" ON)
```
We don't align the different arguments of subsequent function calls

petersteneteg replied:
fixed

alexanderbock commented on CMakeLists.txt:
```diff
@@ -83,11 +86,14 @@ if (UNIX)
 endif ()
 
+#  add dummy uninstall target to avoid it beeing added by dependencies
+if(NOT TARGET uninstall)
```
missing space after `if`

petersteneteg replied:
fixed

alexanderbock commented on ext/CMakeLists.txt:
```diff
@@ -8,7 +8,7 @@
 if (SGCT_DEP_INCLUDE_GLAD) 
   add_subdirectory(glad)
-  set_property(TARGET glad PROPERTY FOLDER "Dependencies")
+  set_target_properties(glad PROPERTIES FOLDER "Dependencies")
```
Neato!  Didn't know about the `set_target_properties` function!

alexanderbock commented on ext/CMakeLists.txt:
```diff
@@ -43,26 +46,29 @@ if (SGCT_DEP_INCLUDE_GLFW)
 
 if (SGCT_DEP_INCLUDE_GLM)
   add_library(glm INTERFACE)
+  add_library(glm::glm ALIAS glm )
```
Extra space before `)`

petersteneteg replied:
fixed

alexanderbock commented on src/apps/stitcher/main.cpp:
```diff
@@ -65,7 +65,7 @@ namespace {
     };
 
-    constexpr std::string_view VertexShader = R"(
+    constexpr std::string_view vertexShader = R"(
```
These should be capitalized as they were before since they are `constexpr`

petersteneteg replied:
ok, needs to change the usage below then,  since that is lowercase...

petersteneteg replied:
modified the usage instead

alexanderbock commented on src/apps/stitcher/main.cpp:
```diff
@@ -79,7 +79,7 @@ void main() {
 )";
 
-    constexpr std::string_view FragmentShader = R"(
+    constexpr std::string_view fragmentShader = R"(
```
Same here

petersteneteg replied:
modified the usage instead

alexanderbock reviewed (changes requested):
Getting:
```
CMake Error at apps/OpenSpace/ext/sgct/src/sgct/CMakeLists.txt:196 (find_package):
  By not providing "Findfmt.cmake" in CMAKE_MODULE_PATH this project has
  asked CMake to find a package configuration file provided by "fmt", but
  CMake did not find one.

  Could not find a package configuration file provided by "fmt" with any of
  the following names:

    fmtConfig.cmake
    fmt-config.cmake

  Add the installation prefix of "fmt" to CMAKE_PREFIX_PATH or set "fmt_DIR"
  to a directory containing one of the above files.  If "fmt" provides a
  separate development package or SDK, be sure it has been installed.
```

when trying to Configure from OpenSpace

petersteneteg committed:
Added install target

There is now an option "SGCT_INSTALL" (off by default) to enable installing SGCT.
Enabled glad from vcpkg.
Address various review comments
Uses taget aliases in all (?) target link libraries, since that will trigger an error if the target is missing where as a "plain" target will just be added textually the the link line.

petersteneteg committed:
Adds a cmake option (SGCT_:BUILD_TESTS=ON) to make it possible to disable building tests

petersteneteg committed:
Only warn about Edit&Continue if we disable it. (formatting fixes)

alexanderbock commented:
Still getting
```
Make Error at apps/OpenSpace/ext/sgct/src/sgct/CMakeLists.txt:184 (find_package):
  By not providing "Findfmt.cmake" in CMAKE_MODULE_PATH this project has
  asked CMake to find a package configuration file provided by "fmt", but
  CMake did not find one.

  Could not find a package configuration file provided by "fmt" with any of
  the following names:

    fmtConfig.cmake
    fmt-config.cmake

  Add the installation prefix of "fmt" to CMAKE_PREFIX_PATH or set "fmt_DIR"
  to a directory containing one of the above files.  If "fmt" provides a
  separate development package or SDK, be sure it has been installed.
```

:(  When compiling in OpenSpace which provides its own FMT target.

Surround with
```
if (NOT TARGET fmt)
...
endif ()
```
?

petersteneteg committed:
Better handling of dependencies when a superpoject is adding targets for us

petersteneteg commented:
I added some handing of the case where a superproject provides targets for sgct to use. It should find fmt now

petersteneteg commented:
Poke @alexanderbock I don't think your jenkins build is configured to build PRs from forks.
